python/pyarrow/ipc.pxi

```diff
@@ -100,6 +100,74 @@ cdef _wrap_read_stats(CIpcReadStats c):
                      c.num_replaced_dictionaries)
 
 
+cdef class IpcReadOptions(_Weakrefable):
+    """
+    Serialization options for reading IPC format.
+
+    Parameters
+    ----------
+    use_threads : bool
+        Whether to use the global CPU thread pool to parallelize any
+        computational tasks like decompression.
+    ensure_native_endian : bool
+        Whether to convert incoming data to platform-native endianness.
+        Default is true.
+    included_fields : list
+        If empty (the default), return all deserialized fields.
+        If non-empty, the values are the indices of fields to read on
+        the top-level schema.
+    memory_pool : MemoryPool, default None
+        Uses default memory pool if not specified
+    """
+    __slots__ = ()
+
+    # cdef block is in lib.pxd
+
+    def __init__(self, *, bint ensure_native_endian=True,
+                 bint use_threads=True, list included_fields=None,
+                 MemoryPool memory_pool=None):
+        self.c_options = CIpcReadOptions.Defaults()
+        self.ensure_native_endian = ensure_native_endian
+        self.use_threads = use_threads
+        self.included_fields = included_fields
+        self.memory_pool = memory_pool
+
+    @property
+    def ensure_native_endian(self):
+        return self.c_options.ensure_native_endian
+
+    @ensure_native_endian.setter
+    def ensure_native_endian(self, bint value):
+        self.c_options.ensure_native_endian = value
+
+    @property
+    def use_threads(self):
+        return self.c_options.use_threads
+
+    @use_threads.setter
+    def use_threads(self, bint value):
+        self.c_options.use_threads = value
+
+    @property
+    def included_fields(self):
+        return self.c_options.included_fields
+
+    @included_fields.setter
+    def included_fields(self, list value):
+        self.c_options.included_fields = value or list()
```

Comment:
We could also use None to signal all columns, instead of an empty list? (like `compression` in the IpcWriteOptions uses None to indicate the default which is a nullptr on the C++ side. Or the default vector is a empty one and not a nullptr?)

Reply:
Hmm vector here isn't nullable so they'd be effectively the same? (We should probably accept both None and empty list though.)

Reply:
Naively in Python, I would interpret an empty list as not reading any columns at all

Reply:
Right. Actually, I think that's a limitation of our current APIs: it's not possible to specify "read no columns"

Reply:
Yes, if we want to use None this way, I think you need to do it like `compression` in the write options, and so do in the init something like:

```
if included_fields is not None:
    self.included_fields = included_fields
```

(so it uses the C++ default if None was specified)

Reply:
my test above was incorrect but my question is the same, is it possible to set `included_fields` to null with the current API implementation?
```python
    def included_fields(self, list value):
        cdef:
            vector[int] included_fields
        if value is None:
            self.c_options.included_fields = included_fields
        else:
            self.c_options.included_fields = value
```
As this still returns an empty list in the case of `None`.

Reply:
vector is not nullable so you can never get None, just an empty list. I think that's fine, but it's just nice to be able to set it to None and do the right thing.

Reply:
Thanks! I have pushed d813029bdf2f4a424e1e267c564acd246cc9aa9b which maintains the default C++ value for `vector[int]` in the case of None.